행렬의 곱

In [ ]:
import numpy as np
a=np.array([[1,2,3],[4,5,6]])
b=np.array([[1,2], [3,4], [5,6]])
np.dot(a,b) # a의 열과 b의 행의 갯수가 같아야지 실행됨

array([[22, 28],
       [49, 64]])

In [ ]:
import numpy as np
a=np.array([[1,2,3],[4,5,6]])
b=np.array([[1,2], [3,4], [5,6]])
c=np.array([[1,2,3,4], [5,6,7,8]])
np.dot(b,c)

array([[11, 14, 17, 20],
       [23, 30, 37, 44],
       [35, 46, 57, 68]])

3층 신경망 구현

In [ ]:
import numpy as np

def sigmoid(x):
    return 1 / (1 + np.exp(-x))

def identity_function(x):
    return x

def init_network():
    network = {}
    network['W1'] = np.array([[0.1, 0.3, 0.5], [0.2, 0.4, 0.6]])
    network['b1'] = np.array([0.1, 0.2, 0.3])
    network['W2'] = np.array([[0.1, 0.4], [0.2, 0.5], [0.3, 0.6]])
    network['b2'] = np.array([0.1, 0.2])
    network['W3'] = np.array([[0.1, 0.3], [0.2, 0.4]])
    network['b3'] = np.array([0.1, 0.2])

    return network
    
def forward(network, x):
    W1, W2, W3 = network['W1'], network['W2'], network['W3']
    b1, b2, b3 = network['b1'], network['b2'], network['b3']

    a1 = np.dot(x, W1) + b1
    z1 = sigmoid(a1)
    a2 = np.dot(z1, W2) + b2
    z2 = sigmoid(a2)
    a3 = np.dot(z2, W3) + b3
    y = identity_function(a3)

    return y
  
network = init_network()
x = np.array([0.1, 0.5])
y = forward(network, x)
print(y)

[0.31234736 0.6863161 ]


소프트맥스 함수

In [ ]:
import numpy as np
def softmax(a):
    c = np.max(a)
    exp_a = np.exp(a-c)
    sum_exp_a = np.sum(exp_a)
    y = exp_a / sum_exp_a

    return y

a = np.array([0.3, 2.9, 4.0])
y = softmax(a)
print(y)

[0.01821127 0.24519181 0.73659691]


소프트맥스 함수의 특징은 모든 수의 합이 1이라는 점이다 따라서 각각 1.8%, 24.5%, 73.7% 로 해석 할 수 있다.

신경망의 추론 처리

In [22]:
import sys, os
sys.path.append('/content/drive/MyDrive/Colab Notebooks/dataset')
from dataset.mnist import load_mnist

def get_data():
    (x_train, t_train), (x_test, t_test) = \
        load_mnist(normalize=True, flatten=True, one_hot_label=False)
    return x_test, t_test

def init_network():
    with open("sample_weight.pkl", 'wb') as f:
        network = pickle.load(f)

    return network

def predict(network, x):
    W1, W2, W3 = network['W1'], network['W2'], network['W3']
    b1, b2, b3 = network['b1'], network['b2'], network['b3']

    a1 = np.dot(x, W1) + b1
    z1 = sigmoid(a1)
    a2 = np.dot(z1, W2) + b2
    z2 = sigmoid(a2)
    a3 = np.dot(z2, W3) + b3
    y = sigmoid(a3)

    return y

x, t = get_data()
network = init_network()

accuracy_cnt = 0
for i in range(len(x)):
    y = predict(network, x[i])
    p = np.argmax(y)
    if p == t[i]:
        accuracy_cnt += 1

print("Accuracy:" + str(float(accuracy_cnt) / len(x)))




ModuleNotFoundError: ignored

오류가 나오는 이유는 지정한 데이터셋이 컴퓨터에 없기 때문입니다.-코랩으로 시도함